chore: Sumcheck audit add tests for lower level functions

kashbrti opened the pull request:
Sumcheck Audit 2 

### 🧾 Audit Context

This PR adds test coverage for the following lower-level functions in sumcheck: 
- `compute_effective_round_size`
- `RowDisablingPolynomials`
- `extend_edges`
- `check_sum` and recursive version of it 
- tempering with evaluations in `proof_then_verify` style flows 

### 🛠️ Changes Made

- added tests mentioned above

kashbrti committed:
add tests for compute_Effective_round_size, extend_edges and row_disabling_polynomials

kashbrti committed:
accumulate_relation_univariate testing

kashbrti committed:
check_sum tests and recursive tests

kashbrti committed:
add evaluation tampering test

kashbrti commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp:
```diff
@@ -708,6 +707,15 @@ template <typename Flavor> class SumcheckProverRound {
         }
     }
 
+    // Methods made accessible for testing
```
This wrapper is annoying but it seemed the easiest way to make the function accessible for the test (I have up on trying to make it work with friend declarations)

iakovenkos replied:
as an alternative - you can introduce a helper in the test fixture that replicates `accumulate_relation_univariates`, but probably it's not as robust if there're any changes to skipping

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck.test.cpp:
```diff
@@ -425,4 +425,130 @@ TYPED_TEST(SumcheckTests, ProverAndVerifierSimpleFailure)
+        }
+
+        // Valid arithmetic circuit: w_l + w_r = w_o (rows 1 and 2)
+        std::array<FF, multivariate_n> w_l = { FF(0), FF(1), FF(2), FF(0), FF(0), FF(0), FF(0), FF(0) };
```
not introduced by you ofc, but I think it's time to make this suite more readable - pls use a helper that populates a couple of arithmetic gates by creating witnesses and then just computing their sums

iakovenkos replied:
though we're bypassing the actual circuit creation in these tests
I guess the real problem here is that we're trying to be somewhat independent of the flavor but at the same time we're explicitly using Ultra in all these tests, so the outcome is somewhere in the middle and is quite confusing. 

Should we create a TestFlavor for all these low-level sanity tests where only the arithmetic relation is present? Then someone reading these tests doesn't need to know anything about UltraFlavor

iakovenkos replied:
+ it can provide more clarity for tests like `extend_edges` as it'd be easy to toggle the modes

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck.test.cpp:
```diff
@@ -425,4 +425,130 @@ TYPED_TEST(SumcheckTests, ProverAndVerifierSimpleFailure)
+
+        auto prover_polynomials = create_valid_circuit();
+
+        RelationParameters<FF> relation_parameters{
```
Looks a bit confusing - z_perm and all sigmas/ids/lagranges are 0 polys, but we're using a specific value for pub inputs delta and use randomness that's only needed for z_perm. Probably should default initialize to 0 with a clarifying comment

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.test.cpp:
```diff
@@ -131,3 +135,922 @@ TEST(SumcheckRound, AddTuplesOfTuplesOfUnivariates)
+        info("Test 4: LogDerivLookupRelation linearly dependent subrelation is not scaled");
+
+        // Create a circuit with lookup-related polynomials
+        std::array<FF, multivariate_n> lookup_read_counts = { FF(1), FF(2), FF(1), FF(0) };
```
This should also be in the TestFlavor and someone reading this shouldn't care if it is log derivative that is linearly dependent or we treat arithmetic relation as a linearly dependent

iakovenkos reviewed (changes requested):
Please add a simple TestFlavor to decouple all the basic tests from Ultra and make the witness creation more readable. It can re-use `UltraCircuitBuilder` as it's the production scenario and then have a simple method to convert the circuit into polys. + some of the modes - like USE_SHORT_MONOMIALS or maybe even subrelations being linearly (in)depenent can be togglable.

kashbrti committed:
Merge branch 'merge-train/barretenberg' into kb/sumcheck_audit_tests

kashbrti committed:
Merge branch 'merge-train/barretenberg' into kb/sumcheck_audit_tests

kashbrti committed:
wip

kashbrti commented:
@iakovenkos I introduced the new flavor `SumcheckTestFlavor`. all tests in `sumcheck.test.cpp` have been migrated to use this flavor. I removed the evaluation tampering test because I noticed we have a `tamper_with_proof` flow that does this outside of sumcheck.

kashbrti commented:
the linear dependence was a pain to do via template parameters (required some meta programming) so I just kept a dependent and independent relation type for all instantiations of the flavor.

kashbrti committed:
remove the broken test

kashbrti committed:
fix the messed up challenge

iakovenkos commented on barretenberg/cpp/src/barretenberg/flavor/sumcheck_test_flavor.hpp:
```diff
@@ -0,0 +1,370 @@
+ * prover_polynomials.q_test = Polynomial({ 1, 1, 1, 1 });  // Activate test relation (dependent)
+ * @endcode
+ */
+template <typename CurveType = curve::BN254, bool HasZK_ = false, bool UseShortMonomials_ = true>
```
nice that you can switch between curves without using ECCVM

kashbrti replied:
Oh actually, I noticed a problem here... I should add this to the `FlavorConcept`. Otherwise, it will never be used.

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck.test.cpp:
```diff
@@ -171,82 +250,18 @@ template <typename Flavor> class SumcheckTests : public ::testing::Test {
-            .public_input_delta = FF::one(),
-        };
+
+        // Use the helper to create a satisfiable trace
```
I think we are trying to avoid such comments in general - the helper name is self-explanatory

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck.test.cpp:
```diff
@@ -388,23 +368,18 @@ template <typename Flavor> class SumcheckTests : public ::testing::Test {
-                                   MegaZKFlavor>;
+// Define the FlavorTypes using SumcheckTestFlavor variants
+// Note: Only testing short monomials since full barycentric adds complexity without testing sumcheck-specific logic
+// Note: Grumpkin does not support ZK sumcheck (LIBRA_UNIVARIATES_LENGTH mismatch)
```
I think we can turn LIBRA_UNIVARIATES_LENGTH assert into a non-strict inequality, when I added it I just wanted to make sure that we don't go over the subgroup size. Maybe the asserts should go to Flavors

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.test.cpp:
```diff
@@ -13,25 +16,28 @@ using namespace bb;
-    Univariate<FF, 2> univariate_2({ 2, 4 });
-    Univariate<FF, 5> univariate_3({ 3, 4, 5, 6, 7 });
+    // SumcheckTestFlavor has: ArithmeticRelation (2 subrelations) + DependentTestRelation (1 subrelation)
+    Univariate<FF, 3> univariate_1({ 1, 2, 3 });       // ArithmeticRelation subrelation 0
```
I'd prefer using `CircuitBuilder` functionality here and a toy "trace_to_polynomials" method to convert the circuit to polynomials, cause this is hard to read

kashbrti replied:
but currently this test is completely abstracted away from Builders etc. I could add another one using builder functionality but I think it's kinda nice that this one is self contained in sumcheck utils.

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.test.cpp:
```diff
@@ -131,3 +141,918 @@ TEST(SumcheckRound, AddTuplesOfTuplesOfUnivariates)
+TEST(SumcheckRound, ExtendEdges)
+{
+    // Use a flavor without ShortMonomials
+    using Flavor = MultilinearBatchingFlavor;
```
switch to the test flavor pls

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.test.cpp:
```diff
@@ -131,3 +141,918 @@ TEST(SumcheckRound, AddTuplesOfTuplesOfUnivariates)
+}
+
+/**
+ * @brief Test accumulate_relation_univariates for UltraFlavor
```
Stale Ultra mention here and in the test name

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.test.cpp:
```diff
@@ -131,3 +141,918 @@ TEST(SumcheckRound, AddTuplesOfTuplesOfUnivariates)
+        typename SumcheckRound::ExtendedEdges extended_edges;
+        round.extend_edges(extended_edges, prover_polynomials, 0);
+
+        typename SumcheckRound::SumcheckTupleOfTuplesOfUnivariates acc1{}, acc2{};
```
please rm multiple declarations in one line

iakovenkos commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.test.cpp:
```diff
@@ -131,3 +141,918 @@ TEST(SumcheckRound, AddTuplesOfTuplesOfUnivariates)
+TEST(SumcheckRound, CheckSumRecursiveUnsatisfiableWitness)
+{
+    using InnerBuilder = bb::UltraCircuitBuilder;
+    using RecursiveFlavor = bb::UltraRecursiveFlavor_<InnerBuilder>;
```
Prob should also add a rec test flavor, especially if you're planning to expand rec sumcheck unit tests

kashbrti replied:
Was thinking about it. But I think it adds sooo much redundancy for not much gain. The recursive logic differs in really few places in `Sumcheck`, so don't see much point in having extensive testing for the recursive version separately (since we already have `UltraRecursive` that is getting tested).

iakovenkos reviewed (approved):
thanks for adding tests for all these low-level methods!

kashbrti committed:
fixing the issue with grumpkin handling and adding the flavor concept.

kashbrti committed:
cleanup

kashbrti committed:
Merge branch 'merge-train/barretenberg' into kb/sumcheck_audit_tests

kashbrti committed:
remove stale comment
